### PR TITLE
Fi 948 migrate profile validation from program

### PR DESCRIPTION
**Pull request by czh-orz**
This PR moves the profile validation tests along with the terminology changes from program to community. I pretty much just copied over the files and whatever other code was needed to make everything run and to pass all the tests. I think special attention needs to be payed to the hl7_validator.rb, terminology.rb, and tasks.rake changes as I'm not really sure if all those changes are necessary/correct.

Before you start testing, make sure you have the newest version of the bloom filters under resources/terminology/validators/bloom.

### CODE REVIEW

**Review by Jammjammjamm — COMMENTED** [review]
I'm concerned about the amount of untested code in this branch, especially considering its complexity.

File | Coverage %
-|-
lib/app/utils/bcp_13.rb | 50
lib/app/utils/bcp47.rb | 29
lib/app/utils/codesystem.rb | 31
lib/app/utils/fhir_package_manager.rb | 35
lib/app/utils/terminology.rb | 51
lib/app/utils/valueset.rb | 52

**Comment by Jammjammjamm** [comment]
I get this error in all of the profile validation tests:

Screen Shot 2020-09-30 at 9 55 56 AM

```
Fatal Error: undefined method `any?' for nil:NilClass
NoMethodError
lib/app/utils/terminology.rb:250:in `validate_code',
lib/app/sequence_base.rb:965:in `block (2 levels) in resources_with_invalid_binding',
lib/app/sequence_base.rb:795:in `block (2 levels) in resolve_element_from_path',
lib/app/sequence_base.rb:784:in `block in resolve_element_from_path',
lib/app/sequence_base.rb:784:in `each',
lib/app/sequence_base.rb:784:in `find',
lib/app/sequence_base.rb:784:in `resolve_element_from_path',
lib/app/sequence_base.rb:795:in `block in resolve_element_from_path',
lib/app/sequence_base.rb:793:in `each',
lib/app/sequence_base.rb:793:in `resolve_element_from_path',
lib/app/sequence_base.rb:936:in `block in resources_with_invalid_binding',
lib/app/sequence_base.rb:932:in `map',
lib/app/sequence_base.rb:932:in `resources_with_invalid_binding',
lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:137:in `block (2 levels) in <class:USCore310PractitionerSequence>'
```

**Comment by Jammjammjamm** [comment]
We may want to make a ticket to look at the performance of this code once it's merged. These changes make the US Core tests take 50% longer for me.

**Comment by czh-orz** [comment]
> I get this error in all of the profile validation tests:
> 
> Screen Shot 2020-09-30 at 9 55 56 AM
> 
> ```
> Fatal Error: undefined method `any?' for nil:NilClass
> NoMethodError
> lib/app/utils/terminology.rb:250:in `validate_code',
> lib/app/sequence_base.rb:965:in `block (2 levels) in resources_with_invalid_binding',
> lib/app/sequence_base.rb:795:in `block (2 levels) in resolve_element_from_path',
> lib/app/sequence_base.rb:784:in `block in resolve_element_from_path',
> lib/app/sequence_base.rb:784:in `each',
> lib/app/sequence_base.rb:784:in `find',
> lib/app/sequence_base.rb:784:in `resolve_element_from_path',
> lib/app/sequence_base.rb:795:in `block in resolve_element_from_path',
> lib/app/sequence_base.rb:793:in `each',
> lib/app/sequence_base.rb:793:in `resolve_element_from_path',
> lib/app/sequence_base.rb:936:in `block in resources_with_invalid_binding',
> lib/app/sequence_base.rb:932:in `map',
> lib/app/sequence_base.rb:932:in `resources_with_invalid_binding',
> lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:137:in `block (2 levels) in <class:USCore310PractitionerSequence>'
> ```

I didn't run into this issue. You may need to update your bloom filters in the resources/terminology folder.

File: lib/app/utils/valueset.rb [hunk 1]
```diff
@@ -156,13 +194,30 @@ def save_csv_to_file(filename = "resources/validators/csv/#{(URI(url).host + URI
         end
       end
 
+      # Load a code system from a file
+      #
+      # @param [String] filename the file containing the code system JSON
+      def self.load_system(filename)
+        # TODO: Generalize this
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/valueset.rb:201]: Not sure about this. @okeefm do you have any thoughts on this?

[the remark this answers is not in the document]
Comment (okeefm) [reply lib/app/utils/valueset.rb:201]: I'm not strictly sure what isn't general enough about this (past-me needed to leave a more useful comment). I think we can probably pull that comment out, and not worry about it until something breaks.

File: lib/app/utils/terminology.rb [hunk 2]
```diff
@@ -1,91 +1,150 @@
...
-          Inferno.logger.debug "Processing #{k}"
-          cs = vs.code_system_set(k)
-          filename = "#{root_dir}/#{bloom_file_name(k)}.msgpack"
-          save_bloom_to_file(cs, filename)
-          validators << { url: k, file: File.basename(filename), count: cs.length, type: 'bloom' }
+      umls_code_systems = Set.new(Inferno::Terminology::ValueSet::SAB.keys)
+      root_dir = "resources/terminology/validators/#{type}"
+      FileUtils.mkdir_p(root_dir)
+
+      get_module_valuesets(selected_module, strengths).each do |k, vs|
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/terminology.rb]: Changed to vs_url

File: lib/app/utils/valueset.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/valueset.rb]: Okay I removed the comment

File: lib/app/utils/bcp47.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          next if language['Type'] == 'region'
+
+          cs_set.add(system: 'urn:ietf:bcp:47', code: language['Subtag']) if meets_filter_criteria?(language, filter)
+        end
+        cs_set
+      end
+
+      private
+
+      @bcp = nil
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/bcp47.rb]: Removed

File: lib/app/utils/bcp47.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        filter_codes(filter)
+      end
+
+      def load_bcp47
+        response = RestClient.get REGISTRY_URL
+        response.body.split(SEPARATOR).drop(1)
+      end
+
+      def filter_codes(filter = nil)
+        bcp47 = bcp
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/bcp47.rb]: Changed to just bcp.each

File: lib/app/utils/bcp_13.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module Inferno
+  class Terminology
+    module BCP13
+      @code_set = nil
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/bcp_13.rb]: Removed

File: lib/app/utils/codesystem.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    class Codesystem
+      attr_accessor :codesystem_model
+
+      def initialize(cs_model)
+        @codesystem_model = cs_model
+      end
+
+      def all_codes_in_concept(concepts)
+        cs_set = Set.new
+        load_codes = lambda do |concept|
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/codesystem.rb]: Removed the lambda

File: lib/app/utils/codesystem.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        starting_concept.each do |cs_concept|
+          return cs_concept if cs_concept.code == concept_code
+
+          next_concepts.push(*cs_concept.concept) unless cs_concept.concept.empty?
+        end
+        next_concepts.each { |next_concept| find_concept(concept_code, next_concept) }
+      end
+
+      def filter_codes(filter = nil)
+        cs_set = nil
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/codesystem.rb]: removed

File: lib/app/utils/fhir_package_manager.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+        path = File.join destination.split('/')
+        FileUtils.mkdir_p(path)
+
+        tar.each do |entry|
+          next if entry.directory?
+
+          next unless entry.full_name.start_with? 'package/'
+
+          file_name = entry.full_name.split('/').last
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/utils/fhir_package_manager.rb]: Fixed

**Review by Jammjammjamm — APPROVED** [review]

**Review by arscan — APPROVED** [review]

**Pull request merged**